### DISCUSSION TITLE
VertexOnlyMesh Improvements Plan

### DISCUSSION THREAD

**Discussion by ReubenHill** [firedrakeproject/firedrake#2450] [General]
The work plan for improvements to VertexOnlyMesh was outlined here https://github.com/firedrakeproject/firedrake/issues/2223.

However, the recent merge of #2437 has made using setting `redundant == PETSC_TRUE` in [DMSwarmSetPointCoordinates](https://petsc.org/release/docs/manualpages/DMSWARM/DMSwarmSetPointCoordinates.html) no longer viable since #2437 switches to using `mesh.locate_cell` for point location.

# First item on old plan

The issue with using `mesh.locate_cell` was outlined in #2224: namely that you get point duplication for points on cell boundaries. The plan had been to update DMSwarmSetPointCoordinates to use `redundant == PETSC_TRUE` by default (meaning that rank 0 would broadcast its point list to all other ranks) such that this would not happen. From #2224:

> Notably it seems that [DMSwarmSetPointCoordinates](https://petsc.org/release/docs/manualpages/DMSWARM/DMSwarmSetPointCoordinates.html) doesn't suffer from this problem so it would be worth checking how the implementation there works.

It seems this isn't correct. Consider `test_parallel_vom_create.py`:
```
from firedrake import *
m = UnitSquareMesh(1,1)
coords = [[0.5, 0.5], [0.9, 0.9]]
vm = VertexOnlyMesh(m, coords)
print(f'Rank {COMM_WORLD.rank}, coordinates {vm.coordinates.dat.data_ro}')
```

Before and after the merging of #2437 running this would give you point duplication:
```
~ mpiexec -n 3 python test_parallel_vom_create.py
Rank 0, coordinates [[0.5 0.5]
 [0.9 0.9]]
Rank 1, coordinates []
Rank 2, coordinates [[0.5 0.5]]
```
Prior to merging #2437 this diff
```
diff --git a/firedrake/mesh.py b/firedrake/mesh.py
index ab8563da..dba3b23d 100644
--- a/firedrake/mesh.py
+++ b/firedrake/mesh.py
@@ -2283,7 +2283,7 @@ def _pic_swarm_in_plex(plex, coords, fields=[]):
     # coordinates associated with them. The DMPlex cell id associated
     # with each PIC in the DMSwarm is accessed with the `DMSwarm_cellid`
     # field.
-    swarm.setPointCoordinates(coords, redundant=False, mode=PETSc.InsertMode.INSERT_VALUES)
+    swarm.setPointCoordinates(coords, redundant=True, mode=PETSc.InsertMode.INSERT_VALUES)
 
     # Remove PICs which have been placed into ghost cells of a distributed DMPlex
     dmcommon.remove_ghosts_pic(swarm, plex)
```
ought to have avoided point duplication (if my comment on #2224 is correct) but, confusingly, it still has the issue
```
~ mpiexec -n 3 python test_parallel_vom_create.py
Rank 0, coordinates [[0.5 0.5]
 [0.9 0.9]]
Rank 1, coordinates []
Rank 2, coordinates [[0.5 0.5]]
```

So what's going on here?
Do I need to do something I've forgotten about to make my diff have an effect?
Was I just wrong when I said it avoided the issue?

# What to do next instead

It's been a while since I wrote the issues and my work plan due to my extended leave of absence. 

Any suggestions on how to proceed?

I will tidy up the various issues above following discussion here.

**Comment by ksagiyam** [comment]
As far as I understand, [DMSwarmSetPointCoordinates](https://petsc.org/release/docs/manualpages/DMSWARM/DMSwarmSetPointCoordinates.html) (followed by [DMLocatePoints](https://petsc.org/release/docs/manualpages/DM/DMLocatePoints.html#DMLocatePoints)) does not have a mechanism to remove such point duplication.

Maybe we have enough PETSc functions exposed to implement the voting algorithm in Firedrake.

**Reply by ReubenHill** [reply]
Looks like my comment
> Notably it seems that [DMSwarmSetPointCoordinates](https://petsc.org/release/docs/manualpages/DMSWARM/DMSwarmSetPointCoordinates.html) doesn't suffer from this problem so it would be worth checking how the implementation there works.

was wrong then. Thanks Koki

**Reply by ReubenHill** [reply]
This is in the process of changing - see https://gitlab.com/petsc/petsc/-/merge_requests/5354 and https://gitlab.com/petsc/petsc/-/merge_requests/5266

**Reply by joewallwork** [reply]
Also https://gitlab.com/petsc/petsc/-/merge_requests/5345

**Comment by dham** [comment]
I think this is all basically fine and nothing we've done really gets in the way.

1. We can move to requiring broadcast from rank 0. That doesn't really change anything else from where we are now. The issues with duplicating or potentially losing points on process ownership boundaries are there now and wouldn't be fixed by this.
2. Given that we're assuming that broadcasting the entire set of points is an OK thing to do, we can fix the process boundary implementation by a simple vote. What we do is change a bit the procedure for dropping vertices we don't own. To do this we would:
   a.  modify the current code so that we attempt to find cells in the halo as well as the core (easy). 
   b. delay actually dropping verticess until after we vote (also easy). We write into number of vertices long integer array the process number we think owns the process. We write -1 if this rank doesn't know. We can then do a collective max over this array to get unambiguous ownership.
   c. now drop the vertices we don't own.

This is all orthogonal to the tolerance issue, which can happen on cell boundaries interior to a process. We should fix that too.

I think we should also change the default behaviour for point loss to raising an error. We shouldn't silently drop vertices unless the user has explicitly indicated that this is OK by them.

**Reply by ReubenHill** [reply]
This is great. 

# API Thoughts
For anyone using `VertexOnlyMesh` at the moment we are proposing default API changes.

There are, I think, 3 variations

1. Broadcasting from rank 0:
```
vm = VertexOnlyMesh(parent_mesh, coords, redundant=True,  point_migration=None)
```
2. Each rank may have different points and points on process boundaries can get duplicated or lost (**current behaviour**)
```
vm = VertexOnlyMesh(parent_mesh, coords, redundant=False, point_migration=False)
```
3. Migrating points to correct rank where each rank has different points (i.e. coming up with rank candidates as outlined in #2223)
```
vm = VertexOnlyMesh(parent_mesh, coords, redundant=False, point_migration=True)
```

I _think_ in most cases people will be using redundant lists of points so (1.) should be the default. 
If `redundant=False` then we should default to case (3.) (`point_migration=True`) but give `NotImplementedError('Point migration is not yet supported, set point_migration to False')` for now forcing it to be explicitly set by the user.

## Error Behaviour

We already have 3 error modes to consider which are already implemented `missing_points_behaviour=None`, `missing_points_behaviour='warn'` and `missing_points_behaviour='error'` which were added in PR #2187.

The existing implementation of this error requires all ranks to advertise the same list of points but the check scales with number of vertices and ranks (see https://github.com/firedrakeproject/firedrake/blob/e6e778ad4745d5d4ceee81722768c2b903ce6a11/firedrake/mesh.py#L2194-L2199) 
With the new proposed implementation I think there must be a way of avoiding this scaling tracking if no process is found for any point but I can't quite work out what it would be. This should also allow us to support case (3.) above.

Thoughts?

**Reply by wence-** [reply]
To determine if any points were dropped, you need to intersect over the processes.

For the case where everyone starts with the same list of vertices, this is trivial given the voting strategy above. Any point whose allocated rank is `-1` after voting was not found anywhere.

If you start with distributed vertices (each rank has its own list of points), then you can compute the intersection as follows:

Determine the set of points I didn't find locally as $M_\text{j} := \{p_i\}$, we have one such set $M_j$ for each rank. Now just compute the intersection over all the ranks of the $\bigcap_{j} M_j$. If this set is not empty, then any points in it were found on no process.

This same strategy also works if you implement point migration for distributed vertices. You just insert a round where you distribute the $M_j$ to those ranks that might have the points, let them search and then you get some new $M_j^\prime$ which you intersect.

**Reply by ReubenHill** [reply]
Re draft API, we [discussed this in the firedrake meeting](https://github.com/firedrakeproject/firedrake/wiki/Firedrake-meeting-2022-05-26) and came to the conclusions
> - DH: Remove point_migration keyword argument as this is premature optimisation.
> - DH: redundant=True should also be strongly encouraged since, if you have so many points as to need it, the locate-cell functionality will kill performance.

**Comment by ReubenHill** [comment]
@dham you mentioned in #2452 that switching to a voting algorithm would allow for points in mesh halos but still ensure uniqueness. How so?

**Reply by dham** [reply]
Consider the following scenario:
Process 0 thinks that the point is just over the edge in its halo and is owned by process 1. It therefore votes that process 1 owns the point.

Process 1, conversely, thinks that the point is just over the edge and is owned by process 0. It therefore votes that process 0 owns the point.

Since both processes think that the point is in their halo, whichever process wins the vote will have a point in its halo, but there will be agreement about who owns that point.

**Reply by ReubenHill** [reply]
Thanks. We will need to update the PETScSF for the VertexOnlyMesh to reflect this by adding leaves as appropriate (see https://github.com/firedrakeproject/firedrake/issues/1793 - I don't yet know how we would do this). Is fixing https://github.com/firedrakeproject/firedrake/issues/2452 the only reason to do this? @ksagiyam ?

**Reply by ksagiyam** [reply]
I think it's something more fundamental: if we want to ensure uniqueness correctly, we need something like the voting algorithm. You could use `PetscSF` or recall what David suggested in the above https://github.com/firedrakeproject/firedrake/discussions/2450#discussioncomment-2815829. Once we have a way to correctly identifying the ownership of each point, there's probably no need to explicitly culling the ghost points, I think.

**Reply by ReubenHill** [reply]
Can you remind me what the point of the halo is on a normal mesh? Voting algorithm aside, why should I need to have points in a parent mesh halo?

**Reply by dham** [reply]
The point of the halo on a normal mesh is to account for the fact that finite element calculations have a stencil. For example if I am assembling an integral, then I need the data on all the entities adjacent to each cell. Since continuous spaces have nodes on entities that are shared between cells (e.g. vertices), and since the ownership boundary has to occur somewhere, it is inevitable that a process will have to compute some cell integrals using a combination of its own data and data from adjacent processors. 

If you now extend this to think about interior facet integrals, these draw on data from two cells. Somewhere you will have to have one of those cells owned by one processor and the other owned by another. 

The point of the halo is that it duplicates data from other processors so that these sorts of operation can occur. The exact depth of the halo required depends on the set of stencil operations supported. The fact that we support interior facet integrals is sufficient to demonstrate that halo cells will exist.

In the case of vertex only data, if we didn't have halo cells then the floating point uncertainty in determining whether a point falls in a cell could result in points "falling through the gaps" between processors. In the case I gave above, both process 0 and process 1 would conclude that they don't own the point and it would be lost. The example I gave above also illustrates why you will get points in halos, independent of the voting algorithm you use.

Now, you could get rid of points in halos again by also implementing vertex migration. In that case, in the scenario above, whichever process ended up owning the point would immediately conclude that it should be sent to the other process. However, that requires more functionality than we have thus far contemplated.

**Reply by ReubenHill** [reply]
So, to combine your and @colinjcotter 's points, vertex only data in halos would never be involved in calculations, but are necessary to ensure every point is owned by a process (if it's in a halo on one process then it must be in an owned, and therefore computed, cell on another process). Am I correct?

> Now, you could get rid of points in halos again by also implementing vertex migration. In that case, in the scenario above, whichever process ended up owning the point would immediately conclude that it should be sent to the other process. However, that requires more functionality than we have thus far contemplated.

I'm not sure I quite understand this without confirming my understanding of the above point. Could you please expand a little?

**Reply by wence-** [reply]
> So, to combine your and @colinjcotter 's points, vertex only data in halos would never be involved in calculations, but are necessary to ensure every point is owned by a process (if it's in a halo on one process then it must be in an owned, and therefore computed, cell on another process). Am I correct?

I think there is some cross-talk.

Suppose you make a vertexonlymesh given some list of points in parallel. Because the _parent mesh_ has halo cells for correct assembly, some of these points may be found in more than one copy of the same cell (on process 0 and process 1 say). Now, only one of those two processes _owns_ that cell, but perhaps the voting algorithm is "lower-ranked process gets the point" and that cell happens to be owned by process 1. So now the vertexonlymesh point `p` is owned on process 0, but the parent mesh cell it lives in is in process 0's halo.

When you interpolate onto a space defined on the vertexonlymesh then you need to evaluation at `p` (which is in process 0's parent mesh halo), but that is fine because you have the data to evaluate things. So it is not true to say "vertex only data in halos is never involved in calculations".

Side note: "to ensure every point is owned by a process" is not possible with this scheme without arbitrary precision floating point, David's scenario is incomplete. Consider the following setup of two cells

```
,----+
|\ 1 |
| \  |
|  x |
| 0 \|
+----`
```
When locating the point `x` on the boundary, cell 0 can answer "it is inside" or "it is outside"; cell 1 can also answer "it is inside" or "it is outside". But these answers are not related to each other. So both cells can happily answer "it is outside" and now `x` doesn't have a cell to go to. So you can "lose" points even without considering parallel meshes.

Now suppose the boundary is actually a boundary between processes (so cell 0 is owned by process 0, but process 0 has cell 1 in its halo, and cell 1 is owned by process 1 but process 1 has cell 0 in its halo). The scenario David sketched is that process 0 says "The point is in cell 1" while process 1 says "The point is in cell 0". So both processes found the point and you need to uniquify by picking a process that will win. If your voting is "I only keep points that are in cells I own" then in this scenario we've lost a point.

But there are some other answers we could have. Process 0 could say "The point is in cell 0 and cell 1" (likewise Process 1), equally Process 0 could say "The point is in neither cell 0 nor cell 1" (likewise process 1). In that latter case, the point disappears.


> > Now, you could get rid of points in halos again by also implementing vertex migration. In that case, in the scenario above, whichever process ended up owning the point would immediately conclude that it should be sent to the other process. However, that requires more functionality than we have thus far contemplated.
> 
> I'm not sure I quite understand this without confirming my understanding of the above point. Could you please expand a little?

Suppose we ignore the horrible corner cases where points go missing. An initial allocation of points to processes ends up with some number of points in halo cells on those processes. At this point we can say "aha, point x is in cell q which is a halo cell on process p, but I know that cell q on process p is owned by process r and is cell t on that process, so let's send the point to process r". After this exchange the vertexonlymesh will (on each process) only contain points that live in owned cells on that process.

**Reply by ReubenHill** [reply]
> When you interpolate onto a space defined on the vertexonlymesh then you need to evaluation at p (which is in process 0's parent mesh halo), but that is fine because you have the data to evaluate things. So it is not true to say "vertex only data in halos is never involved in calculations".

How do you tell process 1, which doesn't have `p` in its _owned_ cell, that process 0 has done the interpolation in its halo? Does that not somehow screw up the interpolation or is there some inter-process communication machinery to deal with that case?

If I've understood correctly, it sounds to me like having VertexOnlyMesh cells in parent mesh halos is unnecessary and confusing: the solution is to (a) use the voting algorithm then (b) migrate as necessary to empty out all halos. Then point evaluation interpolation just happens from owned-to-owned cells.

What have I missed?

> When locating the point x on the boundary, cell 0 can answer "it is inside" or "it is outside"; cell 1 can also answer "it is inside" or "it is outside". But these answers are not related to each other. So both cells can happily answer "it is outside" and now x doesn't have a cell to go to. So you can "lose" points even without considering parallel meshes.

I guess fixing this (other than relying on a user stated tolerance which risks duplication) will need the mesh method `locate_cell` and `locate_cell.c` to have some knowledge of cell candidates. I'm not sure how to do this and `locate_cell.c` has no comments...

**Reply by ReubenHill** [reply]
Note that user stated tolerance was added in #2449

**Reply by wence-** [reply]
> > When you interpolate onto a space defined on the vertexonlymesh then you need to evaluation at p (which is in process 0's parent mesh halo), but that is fine because you have the data to evaluate things. So it is not true to say "vertex only data in halos is never involved in calculations".
> 
> How do you tell process 1, which doesn't have `p` in its _owned_ cell, that process 0 has done the interpolation in its halo? Does that not somehow screw up the interpolation or is there some inter-process communication machinery to deal with that case?

You don't need to. A vertexonlymesh just has some points that are split between processes arbitrarily but in a way which guarantees that we can evaluate at those points using data from the parent mesh that is available locally. So when you interpolate into a space defined on the vertexonlymesh you have all the data you need locally.

> If I've understood correctly, it sounds to me like having VertexOnlyMesh cells in parent mesh halos is unnecessary and confusing: the solution is to (a) use the voting algorithm then (b) migrate as necessary to empty out all halos. Then point evaluation interpolation just happens from owned-to-owned cells.
> 
> What have I missed?

I don't think I understand what is confusing. You need a unique parent cell for each point, you find one (possibly by voting), job done. Whether or not that parent cell happens to be a halo cell on the local process is neither here nor there to further evaluation.

**Reply by wence-** [reply]
> I guess fixing this (other than relying on a user stated tolerance which risks duplication) will need the mesh method `locate_cell` and `locate_cell.c` to have some knowledge of cell candidates. I'm not sure how to do this and `locate_cell.c` has no comments...

The `try_candidate` function (pasted together in pointquery_utils.py) would have to be updated to use robust (rather than inexact) predicates I think, see for example https://www.cs.cmu.edu/~quake/robust.html [there may be more recent work] (AKA, computational geometry is hard enough that just using loose tolerances for this is probably fine).

**Reply by ReubenHill** [reply]
>You don't need to. A vertexonlymesh just has some points that are split between processes arbitrarily but in a way which guarantees that we can evaluate at those points using data from the parent mesh that is available locally. So when you interpolate into a space defined on the vertexonlymesh you have all the data you need locally.

That was a helpful explanation, thank you. 

Assuming that no one can think of any cases where having points in halos is a problem (maybe particle methods?) then I'm satisfied that we can have points in the halos as long as those points are unique (as decided by something like a voting algorithm). We can add point migration later. The PETSc SF for the VertexOnlyMesh will need to be updated - something along the lines of making sure the leaves are correct I think.

**Reply by ReubenHill** [reply]
When we do get to doing point migration, can anyone think why DMSwarmMigrate might not work out of the box if we relabel the DMSWARM with process ownership as necessary?

**Reply by ksagiyam** [reply]
Not sure about `DMSwarmMigrate()`.

After knowing how many particles exist in each cell, I think we can recast the `DMSwarm` as `PetscSection` (with appropriate permutation set as in https://github.com/firedrakeproject/firedrake/blob/93e1c643d9d003aa20d384bc163fd97300f1ea57/firedrake/cython/dmcommon.pyx#L948) representing the non-uniform DoF layout over the cell stratum. Once this is done, I think we can use  `PetscSFCreateSectionSF()` with the parent `plex`'s `pointSF` to get `swarm`'s `pointSF`.

Then I can directly use this `PetscSection` to checkpoint a `VertexOnlyMesh` and the fields defined on it.

**Reply by ReubenHill** [reply]
> recast the `DMSwarm` as `PetscSection`

What do you mean by this? All I think I know about `PetscSection` is that we use them to hold certain bits of function space information. A DMSwarm, here taking a similar role to a DMPlex in other meshes, seems to me to be a very different data type

**Reply by dham** [reply]
@ksagiyam  wants to attach the swarm to the plex of the parent mesh in order to checkpoint it. He observes that you can think of a swarm as a function space over the parent mesh in which there are varying numbers of degrees of freedom in each cell. This effectively makes a way to store the swarm in the plex. It's probably not what you want in memory when you're computing (because a swarm is the right abstraction) but it's a feasible way of checkpointing and loading a swarm and the data on it.

**Comment by colinjcotter** [comment]
It is for computing values for points that depend on values on points that belong to other processors.

It seems like you don't need to have points in a parent mesh halo, because another processor will compute them.
________________________________

**Comment by danshapero** [comment]
Are there possible improvements to single processor performance from using a different spatial data structure for point location? I'm not familiar with how a lot of the parallelism works under the hood, so take this with a grain of salt, but there could be a lot of value to be gained for less effort from putting bounding boxes for each triangle into an R-tree or quadtree.

**Reply by wence-** [reply]
we do that already
